### PR TITLE
Fix: do not extract/lint html by default

### LINKED ISSUE #1828: disallowTabs fails with TypeError, Extracting HTML by default
```
TypeError: Cannot call method 'replace' of undefined
    at renderLine (/home/vagrant/www/node_modules/jscs/lib/errors.js:250:17)
    at Object.Errors.explainError (/home/vagrant/www/node_modules/jscs/lib/errors.js:181:13)
    at /home/vagrant/www/node_modules/jscs/lib/reporters/console.js:16:36
    at Array.forEach (native)
    at /home/vagrant/www/node_modules/jscs/lib/reporters/console.js:14:35
    at Array.forEach (native)
    at Object.module.exports [as writer] (/home/vagrant/www/node_modules/jscs/lib/reporters/console.js:9:22)
    at /home/vagrant/www/node_modules/jscs/lib/cli.js:151:18
    at Array.0 (/home/vagrant/www/node_modules/vow/lib/vow.js:711:56)
    at Object.callFns [as _onImmediate] (/home/vagrant/www/node_modules/vow/lib/vow.js:23:35)
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27081681-disallowtabs-fails-with-typeerror-extracting-html-by-default?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by hzoo**
Fixes #1828
Closes gh-1870

### CODE REVIEW

**Comment by hzoo** [comment]
@lahmatiy turns off extract by default and removes boolean option.

File: test/specs/checker.js [hunk 1]
```diff
@@ -177,16 +173,6 @@ describe('checker', function() {
             });
         });
 
-        it('extractFile should return null if file doesn\'t check', function() {
-            checker.configure({
-                extract: false
```
Comment (lahmatiy) [inline test/specs/checker.js]: this test should be ok, just remove `checker.configure`

File: lib/config/configuration.js [hunk 2]
```diff
@@ -489,11 +489,8 @@ Configuration.prototype._processConfig = function(config) {
     }
 
     if (options.hasOwnProperty('extract')) {
-        this._loadExtract(options.extract);
-
-    // Set default masks if there is no presets that could define their own
-    } else if (!options.hasOwnProperty('preset')) {
-        this._loadExtract(this._defaultExtractFileMasks);
+        assert(Array.isArray(options.extract), '`extract` option requires array value');
+        this._loadExtract(options.extract.concat(['**/*.+(htm|html|xhtml)']));
```
Comment (lahmatiy) [inline lib/config/configuration.js]: It's wrong. Should pass `options.extract` as is, with no extra.

Comment (lahmatiy) [inline lib/config/configuration.js]: Moreover, this mask stores in `_defaultExtractFileMasks`

File: lib/config/configuration.js [hunk 3]
```diff
@@ -745,11 +742,7 @@ Configuration.prototype._loadExcludedFiles = function(masks) {
  * @protected
  */
 Configuration.prototype._loadExtract = function(masks) {
-    if (masks === false) {
```
Comment (lahmatiy) [inline lib/config/configuration.js]: I think should be possible to pass `true` value to use default mask for html files.

**Comment by lahmatiy** [comment]
> Default value is ["**/*.+(htm|html|xhtml)"]

Also should be removed from docs

File: test/specs/extract-js.js [hunk 4]
```diff
@@ -38,7 +38,8 @@ describe('extract-js', function() {
     describe('whitespaces', function() {
         it('should be no errors on disallowTrailingWhitespace', function() {
             checker.configure({
-                disallowTrailingWhitespace: true
+                disallowTrailingWhitespace: true,
+                extract: []
```
Comment (lahmatiy) [inline test/specs/extract-js.js]: Looks strange. `extract: true` would be better

**Comment by hzoo** [comment]
If they specify an array at all - would we want to add `.html`?
Although If we only support html anyway might as well only allow true

**Comment by hzoo** [comment]
@lahmatiy ok made changes

**Comment by lahmatiy** [comment]
> Although If we only support html anyway might as well only allow true

Nope. For example your project can has templates or other html file with extensions other than defaults.

File: lib/config/configuration.js [hunk 5]
```diff
@@ -745,12 +741,15 @@ Configuration.prototype._loadExcludedFiles = function(masks) {
...
+    if (masks === true) {
+        masks = this._defaultExtractFileMasks;
+    } else if (masks === false) {
         masks = [];
+    } else {
+        assert(Array.isArray(masks), '`extract` option should be array of strings');
+        masks = masks.concat(this._defaultExtractFileMasks);
```
Comment (lahmatiy) [inline lib/config/configuration.js]: You still concat default masks. It's not a good idea. Left users to specify exact set of masks.

Comment (hzoo) [inline lib/config/configuration.js]: So you're saying we should only use the defaults for value `true`?

File: OVERVIEW.md [hunk 6]
```diff
@@ -313,7 +313,7 @@ JavaScript extracting from files that doesn't match to [fileExtensions](#fileExt
 
 #### Default
 
-Default value is `["**/*.+(htm|html|xhtml)"]`. So JavaScript is extracted from files with `.htm`, `.html` or `.xhtml` extension.
+JavaScript is extracted from files with `.htm`, `.html` or `.xhtml` extension by default. If an array is passed those extensions are concatenated.
```
Comment (lahmatiy) [inline OVERVIEW.md]: It's not correct. By default it doesn't extract anything.
Array contains file masks, not extensions.
`true` value is use default html mask.

File: lib/config/configuration.js (hunk 5 above)
Comment (lahmatiy) [inline lib/config/configuration.js]: Yes, exactly.

**Comment by lahmatiy** [comment]
> we only support html

It's for now. But I believe someday it will be expanded for other formats.

**Comment by hzoo** [comment]
@lahmatiy ok changes made

File: test/specs/config/configuration.js [hunk 7]
```diff
@@ -354,21 +354,39 @@ describe('config/configuration', function() {
     });
 
     describe('shouldExtractFile', function() {
-        it('should be check *.htm, *.html, *.xhtml by default', function() {
+        it('should not check anything by default', function() {
             configuration.load({});
+            expect(!!configuration.shouldExtractFile('file.htm')).to.equal(false);
```
Comment (lahmatiy) [inline test/specs/config/configuration.js]: `shouldExtractFile` always returns boolean, maybe remove `!!` and `!`?

**Comment by lahmatiy** [comment]
Looks good :+1:

File: test/specs/config/configuration.js (hunk 7 above)
Comment (hzoo) [inline test/specs/config/configuration.js]: yeah otherwise looks confusing

**Pull request merged**